velox/substrait/VeloxToSubstraitExpr.cpp

```diff
@@ -0,0 +1,452 @@
+
+#include "velox/expression/ControlExpr.h"
+
+namespace facebook::velox::substrait {
+
+const ::substrait::Expression& VeloxToSubstraitExprConvertor::toSubstraitExpr(
+    google::protobuf::Arena& arena,
+    const std::shared_ptr<const ITypedExpr>& expr,
+    const RowTypePtr& inputType) {
+  ::substrait::Expression* substraitExpr =
+      google::protobuf::Arena::CreateMessage<::substrait::Expression>(&arena);
+  if (auto constExpr =
+          std::dynamic_pointer_cast<const ConstantTypedExpr>(expr)) {
+    substraitExpr->mutable_literal()->MergeFrom(
+        toSubstraitExpr(arena, constExpr));
+    return *substraitExpr;
+  }
+  if (auto callTypeExpr =
+          std::dynamic_pointer_cast<const CallTypedExpr>(expr)) {
+    substraitExpr->MergeFrom(toSubstraitExpr(arena, callTypeExpr, inputType));
+    return *substraitExpr;
+  }
+  if (auto fieldExpr =
+          std::dynamic_pointer_cast<const FieldAccessTypedExpr>(expr)) {
+    substraitExpr->mutable_selection()->MergeFrom(
+        toSubstraitExpr(arena, fieldExpr, inputType));
+
+    return *substraitExpr;
+  }
+  if (auto castExpr = std::dynamic_pointer_cast<const CastTypedExpr>(expr)) {
+    substraitExpr->mutable_cast()->MergeFrom(
+        toSubstraitExpr(arena, castExpr, inputType));
+
+    return *substraitExpr;
+  } else {
+    VELOX_UNSUPPORTED("Unsupport Expr '{}' in Substrait", expr->toString());
+  }
+}
+
+const ::substrait::Expression_Cast&
+VeloxToSubstraitExprConvertor::toSubstraitExpr(
+    google::protobuf::Arena& arena,
+    const std::shared_ptr<const CastTypedExpr>& castExpr,
+    const RowTypePtr& inputType) {
+  ::substrait::Expression_Cast* substraitCastExpr =
+      google::protobuf::Arena::CreateMessage<::substrait::Expression_Cast>(
+          &arena);
+  std::vector<std::shared_ptr<const ITypedExpr>> castExprInputs =
+      castExpr->inputs();
+
+  substraitCastExpr->mutable_type()->MergeFrom(
+      typeConvertor_.toSubstraitType(arena, castExpr->type()));
+
+  for (auto& arg : castExprInputs) {
+    substraitCastExpr->mutable_input()->MergeFrom(
+        toSubstraitExpr(arena, arg, inputType));
+  }
+  return *substraitCastExpr;
+}
+
+const ::substrait::Expression_FieldReference&
+VeloxToSubstraitExprConvertor::toSubstraitExpr(
+    google::protobuf::Arena& arena,
+    const std::shared_ptr<const FieldAccessTypedExpr>& fieldExpr,
+    const RowTypePtr& inputType) {
+  ::substrait::Expression_FieldReference* substraitFieldExpr =
+      google::protobuf::Arena::CreateMessage<
+          ::substrait::Expression_FieldReference>(&arena);
+
+  const auto& exprType = fieldExpr->type();
+  std::string exprName = fieldExpr->name();
+
+  ::substrait::Expression_ReferenceSegment_StructField* directStruct =
+      substraitFieldExpr->mutable_direct_reference()->mutable_struct_field();
+
+  const auto& preNodeColNames = inputType->names();
+  const auto& preNodeColTypes = inputType->children();
+  const auto& preNodeColNums = preNodeColNames.size();
```

Comment:
This variable is not needed and there is not point in using a reference for an integer type.

Reply:
remove them.